### ISSUE TITLE
Expose `Utf8LossyChunksIter`

### ISSUE THREAD

**Issue by dylni** [rust-lang/libs-team#54]
# Proposal

## Problem statement

When [`Utf8Error::valid_up_to`] and [`Utf8Error::error_len`] are used, their results will almost always be used to get substrings of the original string. However, since [`Utf8Error`] does not have a reference to the original string, it cannot have methods to return the substrings.

[`Utf8LossyChunksIter`] is also much easier to use.

## Motivation, use-cases

This is useful when creating a custom byte string formatter. UTF-8 portions are usually output using the [`Display` implementation for `str`] or [`str::escape_debug`], but invalid portions might require custom formatting.

### Example

Code using [`str::from_utf8`] (requires `unsafe`):

```rust
while !string.is_empty() {
    let (valid, invalid) = match str::from_utf8(string) {
        Ok(string) => (string, &[][..]),
        Err(error) => {
            let valid_len = error.valid_up_to();
            let valid = unsafe { str::from_utf8_unchecked(&string[..valid_len]) };
            let mut invalid = &string[valid_len..];
            if let Some(invalid_len) = error.error_len() {
                invalid = &invalid[..invalid_len];
            }
            (valid, invalid)
        }
    };
    // formatting for `valid` and `invalid`
    string = &string[valid.len() + invalid.len()..];
}
```

Code using the new API:

```rust
for chunk in Utf8Chunks::new(string) {
    let valid = chunk.valid();
    let invalid = chunk.invalid();
    // formatting for `valid` and `invalid`
}
```

## Solution sketches

Make the following changes, and change the feature for these structs from `str_internals` to `utf8_chunks`.
- Remove [`Utf8Lossy`].
- Rename [`Utf8LossyChunksIter`] to `Utf8Chunks`.
- Rename [`Utf8LossyChunk`] to `Utf8Chunk`.
- Rename [`Utf8LossyChunk::broken`] to `invalid`.
- Change fields of `Utf8Chunks` into getter methods (i.e., `valid` and `invalid`).
- Add:
    ```rust
    impl<'a> Utf8Chunks<'a> {
        fn new(bytes: &'a [u8]) -> Self;
    }
    ```

## What happens now?

This issue is part of the libs-api team [API change proposal process]. Once this issue is filed the libs-api team will review open proposals in its weekly meeting. You should receive feedback within a week or two.

[API change proposal process]: https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html

[`Display` implementation for `str`]: https://doc.rust-lang.org/core/primitive.str.html#impl-Display
[`str::escape_debug`]: https://doc.rust-lang.org/std/primitive.str.html#method.escape_debug
[`str::from_utf8`]: https://doc.rust-lang.org/std/str/fn.from_utf8.html
[`Utf8Error`]: https://doc.rust-lang.org/std/str/struct.Utf8Error.html
[`Utf8Error::error_len`]: https://doc.rust-lang.org/std/str/struct.Utf8Error.html#method.error_len
[`Utf8Error::valid_up_to`]: https://doc.rust-lang.org/std/str/struct.Utf8Error.html#method.valid_up_to
[`Utf8Lossy`]: https://doc.rust-lang.org/core/str/lossy/struct.Utf8Lossy.html
[`Utf8LossyChunk`]: https://doc.rust-lang.org/core/str/lossy/struct.Utf8LossyChunk.html
[`Utf8LossyChunk::broken`]: https://doc.rust-lang.org/core/str/lossy/struct.Utf8LossyChunk.html#structfield.broken
[`Utf8LossyChunksIter`]: https://doc.rust-lang.org/core/str/lossy/struct.Utf8LossyChunksIter.html

**Comment by m-ou-se** [comment]
The `bstr` crate has a similar API, which also exposes another field/method on the chunks: `incomplete`, to check whether the incomplete sequence might still become valid if more bytes are added: https://docs.rs/bstr/latest/bstr/struct.Utf8Chunk.html

**Comment by m-ou-se** [comment]
Regardless of that minor detail, I'm in favor of working towards exposing this as a public API!

**Comment by dylni** [comment]
Great! I agree that basing the API off bstr would be sensible, so I have updated the issue. The changes were actually surprisingly close already, but providing methods instead of fields will be better for future compatibility.

I'm leaving out the `incomplete` method for now, since that can be added later and I don't personally need it. However, I can add it back if it seems useful.

**Comment by dylni** [comment]
@m-ou-se Is this proposal ready for a PR? It still has the "initial-comment-period" label, so I'm not sure if it's been agreed that this should be added.

**Comment by m-ou-se** [comment]
@dylni Sorry! We don't have this processs automated yet. Yes, feel free to go ahead and [open a tracking issue](https://github.com/rust-lang/rust/issues/new?assignees=&labels=C-tracking-issue%2C+T-libs&template=library_tracking_issue.md&title=Tracking+Issue+for+XXX) for this new unstable feature and send a PR for its implementation.

**Comment by dylni** [comment]
Thanks!

Tracking issue: rust-lang/rust#99543

**Comment by dylni** [comment]
This has been merged: rust-lang/rust#99544
